src/main/java/com/iota/iri/service/tipselection/impl/ReferenceCheckerImpl.java

```diff
@@ -0,0 +1,62 @@
+package com.iota.iri.service.tipselection.impl;
+
+import java.util.Arrays;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.function.Predicate;
+
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.service.tipselection.ReferenceChecker;
+import com.iota.iri.storage.Tangle;
+
+import com.iota.iri.utils.dag.DAGHelper;
+import org.apache.commons.collections4.CollectionUtils;
+
+public class ReferenceCheckerImpl implements ReferenceChecker {
+    private Tangle tangle;
+
+    public ReferenceCheckerImpl(Tangle tangle) {
+        this.tangle = tangle;
+    }
+
+    @Override
+    public boolean doesReference(Hash referencer, Hash target) throws Exception {
+        if (!TransactionViewModel.exists(tangle, referencer) ||
+            !TransactionViewModel.exists(tangle, target)) {
+                throw new IllegalStateException("Transactions not found");
+        }
+
+        LinkedHashSet<Hash> visitedHashes = new LinkedHashSet<>();

```

Comment:
Is there a specific reason for LinkedHashSet over HashSet as no iteration is ever performed on it?